### ISSUE TITLE
A widespread but rare race condition with Turbo Stream connections

### ISSUE THREAD

**Issue by keithschacht** [rails/rails#52420]
Long ago, Rails fixed a race condition within ActiveRecord with find_or_create by introducing a create_or_find. I found a somewhat similar race condition with Turbo Stream.

Basically, anytime a `turbo_stream_from` connection is established on a page, the contents can immediately be out of date because of the tiny gap between fetching the record and establishing the channel connection to listen for updates. To elaborate:

### Steps to reproduce

1. A `show` page loads, the controller does a simple Record.find(), and the view displays the object. This view has a `turbo_stream_from @record` in case there are any updates to this record while the user has this page open.
2. However, in the very short period between executing `Record.find()` and the turbo stream connection being established, if the server pushes a `record.broadcast_*` it is missed by the client. The result is that the show page loads and it's already out of date. It's not showing the latest version of the record, and it won't update again until an additional change to the record gets broadcasted out.

It's difficult to write a test script that consistently executes the race condition but here is an example of the log output from a instance where the race was hit:

```
web      | Started POST "/assistants/1/messages" for 23.112.39.144 at 2024-07-25 16:27:54 -0500
...
web      |   Message Create (3.8ms)  INSERT INTO "messages" ...
web      |   ↳ app/controllers/messages_controller.rb:43:in `create'
web      | Redirected to https://krschacht-gmail-com.ngrok.io/conversations/177/messages
web      | Completed 303 See Other in 88ms (ActiveRecord: 43.2ms | Allocations: 48958)
web      | 
worker   | [SolidQueue] Claimed 1 jobs
web      | Started GET "/conversations/177/messages" for 23.112.39.144 at 2024-07-25 16:27:54 -0500
web      | Processing by MessagesController#index as TURBO_STREAM
worker   | 
worker   | ### MessageUpdateJob.perform(1, 954, 1, 1)
web      |   Rendering layout layouts/application.html.erb
web      |   Rendering messages/index.html.erb within layouts/application
web      |   Rendered layout layouts/application.html.erb (Duration: 113.3ms | Allocations: 177929)
web      | Completed 200 OK in 137ms (Views: 102.6ms | ActiveRecord: 14.7ms | Allocations: 189824)
web      | 
web      | [ActionCable] Broadcasting to Z2lkOi8vaG9zdGVkLWdwdC9Db252ZXJzYXRpb24vMTc3: "<turbo-stream action=\"update\" target=\"message_954\"><template> ...   | [SolidQueue] Claimed 1 jobs
worker   | ### Finished MessageUpdateJob.perform(1, 954, 1, 1)
...
web      | Started GET "/cable" for 23.112.39.144 at 2024-07-25 16:27:57 -0500
web      | Started GET "/cable" [WebSocket] for 23.112.39.144 at 2024-07-25 16:27:57 -0500
web      | Successfully upgraded to WebSocket (REQUEST_METHOD: GET, HTTP_CONNECTION: Upgrade, HTTP_UPGRADE: websocket)
web      | Turbo::StreamsChannel is transmitting the subscription confirmation
#
# Channel is now connected but it missed the ActionCable broadcast which happened right before it
#
```

### Expected behavior

If a controller/view fetches a record and then establishes a stream connection to listen for updates to that record, I expect it not to miss any stream updates.

### System configuration

**Rails version**: rails (7.1.3.2)

**Ruby version**: ruby 3.2.3

### Possible Solution

I implemented a work-around within my app. Specifically, I listen for when a Turbo::StreamChannel is established and immediately re-broadcast the record after it's connected. Unfortunately this fix was not trivial to implement because the connection notification which can be subscribed to is missing the identifier. I tee'd up PR https://github.com/rails/rails/pull/52421 to correct the notification.

With the additional details on that notification I implemented this solution to the race condition:

```
module ApplicationCable
  class Connection < ActionCable::Connection::Base
    def connect
      ActiveSupport::Notifications.subscribe "transmit_subscription_confirmation.action_cable" do |*args|
        # Sometimes the conversation page loads and it doesn't get the message to appear, even though
        # the logs show it was broadcasted. This is because there is a small race condition where right
        # in between when find() of the message and the client establishing the turbo stream connection,
        # the broadcast happens.
        #
        # The fix is to re-broadcast the last message to the client after the connection is established.
        event = ActiveSupport::Notifications::Event.new(*args)

        if event.payload[:channel_class] == "Turbo::StreamsChannel"
          identifier = JSON.parse(event.payload[:identifier])
          signed_stream_name = identifier["signed_stream_name"]
          stream_name = Base64.urlsafe_decode64(Turbo.signed_stream_verifier.verify(signed_stream_name))
          class_name, id = stream_name.split("/").last(2)
          conversation = class_name.classify.constantize.find(id)
          message = conversation.messages.ordered.last

          MessageUpdateJob.rebroadcast_message(message)
        end
      end
    end
  end
end
```

Is this something we should generalize and fix within Turbo Stream or ActionCable?

**Comment by josh-m-sharpe** [comment]
Yes. 100% @rails-bot reopen this.

**Comment by josh-m-sharpe** [comment]
I have a relatively simple setup:

<%= turbo_stream_from 'foo' %>
 -> SomeJob.perform_later
   -> Turbo::StreamsChannel.broadcast_replace_to(...)

This results in the broadcast happening before the client connects to `/cable` therefore missing the broadcast.

I would expect this setup to work out of the box.

**Comment by yaroslavrick** [comment]
Yes, I have same issue

**Comment by sfnelson** [comment]
@rails-bot  this is still an issue in `8-0-stable`. Reproductions provided are sufficient.

**Comment by keithschacht** [comment]
@sfnelson Thanks for chiming in on this. I was just going to give up when I saw this rails bot comment yesterday. :) When I open a ticket and it gets auto-closed with no attention, I generally just give up and figure no one really cares. But your note and the others chiming in reminds me: I'm not the only one and it's worth getting to the bottom of the issue with a resolution. I was even circling around one possible solution with my ActiveSupport subscribe hook. I also know that AnyCables (by evil martins) has somehow solved this, so it could be worth someone investigating what the solution is to that. Someone could ping the engineers on that project to ask them.

**Comment by sfnelson** [comment]
@krschacht I think this is "working as intended" for some definitions of intended. The time taken to establish a connection is very low, and if your work is going to complete that fast then an async job + broadcast is possibly not the right tool. I doubt it going to get traction from the core team, but personally find this sort of issue helpful for community members to understand the rough edges of the Rails framework.

That said, there are cases where your job might complete immediately or take a long time, and in that case having a clearly-explained and detailed approach would be immensely valuable.

**Comment by keithschacht** [comment]
@sfnelson I agree with your assessment. Just to capture it on this thread, the other case this came up for me was long-lived connections. I added some logging to detect disconnect and reconnects of the channel and they do regularly happen for reasons I couldn't identify and sometimes broadcasts were missed in those situations too so it's not just quickly completing jobs.

**Comment by kinduff** [comment]
Solved this using a polling strategy in the frontend using Javascript. Each widget has a StimulusJS controller connected, and will check for an internal state to track changes and its completed status. It will update every X seconds (with a max attempts/time) until the background job completely process the resource. If the bug happens, it will just update once more in the first poll avoiding this issue entirely.

**Comment by bhtabor** [comment]
Here is a work around using stimulus and turbo frame. It basically delays the request from a turbo frame until the WebSocket is connected (by using turbo frame `data-src` instead of `src`).

```
cat app/javascript/controllers/core/turbo_stream_frame_controller.js

import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["frame", "stream"]

  connect() {
    if (this.hasStreamTarget) {
      this.observe()
    }
  }

  disconnect() {
    this.observer?.disconnect()
  }

  observe() {
    this.observer = new MutationObserver(() => {
      if (this.streamTarget.hasAttribute("connected")) {
        this.loadFrame()
      }
    })

    this.observer.observe(this.streamTarget, { attributeFilter: ["connected"] })
  }

  loadFrame() {
    if (this.hasFrameTarget && this.frameTarget.dataset.src) {
      this.frameTarget.src = this.frameTarget.dataset.src
    }
  }
}

```

**Comment by sfnelson** [comment]
@rails-bot Still relevant. Please keep this issue open
